### ISSUE TITLE
Approve Race Condtion

### ISSUE THREAD

**Issue by GalloDaSballo** [code-423n4/org#51]

**Comment by GalloDaSballo** [comment]
Other Judges (Please add to this):

@0xean 
Initially High
https://github.com/code-423n4/2022-08-olympus-findings/issues/77 

Changed to QA
https://github.com/code-423n4/2022-10-thegraph-findings/issues/293

**Comment by pashov** [comment]
I think the explanation of the vulnerability is missing to show it's actual problem. Here is an example scenario:

1. Bob approves Alice to spend 100 of his tokens
2. After a while, Bob decides Alice should actually be able to spend only 50 of his tokens and sends a tx to set the allowance to 50
3. Alice sees this tx and front-runs it to spend 100 tokens
4. Now Alice again has an allowance of 50 and she can transfer those tokens out, totalling 150 tokens transferred

So Bob has only approved Alice to spend 100 and then decreased to 50, but she was able to transfer 150. Yes, you could say when Bob approved Alice to spend 100 tokens he could have treated the tokens as spent/lost, but then he lost 50 more.

Essentially this is stealing value from a user which should result in at least a Medium severity vulnerability in my opinion.

**Comment by trust1995** [comment]
I agree "race conditions area property of public mempool chains", however there are good ways to mitigate this issue at the contract level so that the specific risk of losing X+Y of approve(X), approve(Y) is removed. Since 90% of users, I would say, do not expect that they are trusting spender  for X+Y, but for MAX(X,Y), it is worthy enough to point out as M risk for sponsors.

For those unfamiliar with double spend mitigation, the options are:
1. [increaseAllowance](https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#ERC20-increaseAllowance-address-uint256-)() and decreaseAllowance().
2. Force approves to toggle between zero and non-zero value

This is _not_ to say I don't agree that users should consider any approved tokens as spent due to frontrunning risks.

**Comment by GalloDaSballo** [comment]
> I agree "race conditions area property of public mempool chains", however there are good ways to mitigate this issue at the contract level so that the specific risk of losing X+Y of approve(X), approve(Y) is removed. Since 90% of users, I would say, do not expect that they are trusting spender for X+Y, but for MAX(X,Y), it is worthy enough to point out as M risk for sponsors.
> 
> For those unfamiliar with double spend mitigation, the options are:
> 
> 1. [increaseAllowance](https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#ERC20-increaseAllowance-address-uint256-)() and decreaseAllowance().
> 2. Force approves to toggle between zero and non-zero value
> 
> This is _not_ to say I don't agree that users should consider any approved tokens as spent due to frontrunning risks.

What severity would you recommend given your considerations?

**Comment by trust1995** [comment]
Unless it is an issue we have evidence that sponsor is aware of, I would say M.

**Comment by GalloDaSballo** [comment]
Additional counter argument against Med is, why isn't address(0) also a Med then?

Answer, it's not med because it's the user that self-directs the "odd behaviour", in the same way approving non-zero to non-zero, with a target that is malicious, is an odd behaviour, if our severities have any consistency then we cannot argue Med to be appropriate even if you believe this is a valid vulnerability due to the triple conditionality:
- Reckless User
- Non-Zero to non-zero implementation
- Malicious attacker that can front-run

**Comment by trust1995** [comment]
We can look at it as different threat scenarios, frontrunning risk of first order, and of second order.
approve(100) and then approve(0) - users are mostly aware that their 100 is at risk
approve(100) and then approve(150) - users are aware that 150 is at risk, not that 250 is at risk.

Concerning your reasoning:
If we have a majority that agrees this behavior is only exhibited by reckless users, then I would definitely agree to decreasing it to QA level.  That's why it's important to get some more opinions on this.

**Comment by pashov** [comment]
> 

Allowance is a good functionality and a great idea. The problem is in the way ERC20's `approve()` function sets it, just overwriting the old value in storage. If for example `decreaseAllowance()` is used, here is how my previous example looks:

1. Bob approves Alice to spend 100 of his tokens
2. After a while, Bob decides Alice should actually be able to spend only 50 of his tokens and sends a tx to **decrease** the allowance by 50
3. Alice sees this tx and front-runs it to spend 100 tokens
4. Now Bob's tx to decrease the allowance fails, because it tries to do the math `0 - 50` which will underflow and throw an error.

This way Alice could have spent a maximum of 100 tokens, instead of 150 - Bob is protected.

And for the zero address checks, I think this is a great resource - https://forum.openzeppelin.com/t/removing-address-0x0-checks-from-openzeppelin-contracts/2222

I'd say zero-address checks vulnerability is not that a user is inputting 0 by mistake, but if the call was "coded" incorrectly, missing an argument or something. This has been an issue in older Solidity versions but since now there are msg.data length validations, this is a much lower severity issue.

**Comment by Picodes** [comment]
I also think this should be QA. In addition to what @GalloDaSballo pointed out, the attack also requires that the attacker waits until he sees an approval transaction in the mempool to act and extract the user's funds, which seems unlikely to me: the user could transfer its tokens, use a private rpc to change its allowance, etc. The most rational thing for the attacker is to exploit the approval asap.

**Comment by GalloDaSballo** [comment]
> > 
> 
> Allowance is a good functionality and a great idea. The problem is in the way ERC20's `approve()` function sets it, just overwriting the old value in storage. If for example `decreaseAllowance()` is used, here is how my previous example looks:
> 
> 1. Bob approves Alice to spend 100 of his tokens
> 2. After a while, Bob decides Alice should actually be able to spend only 50 of his tokens and sends a tx to **decrease** the allowance by 50
> 3. Alice sees this tx and front-runs it to spend 100 tokens
> 4. Now Bob's tx to decrease the allowance fails, because it tries to do the math `0 - 50` which will underflow and throw an error.
> 
> This way Alice could have spent a maximum of 100 tokens, instead of 150 - Bob is protected.
> 
> And for the zero address checks, I think this is a great resource - https://forum.openzeppelin.com/t/removing-address-0x0-checks-from-openzeppelin-contracts/2222
> 
> I'd say zero-address checks vulnerability is not that a user is inputting 0 by mistake, but if the call was "coded" incorrectly, missing an argument or something. This has been an issue in older Solidity versions but since now there are msg.data length validations, this is a much lower severity issue.

The example shows how this is ultimately a biased discussion (half-full, half-empty glass), you are claiming that decreaseAllowance reduces the total at risk to 100, I'm claiming that the 100 is at risk because of the initial allowance and will act accordingly.

There's a point at which personal responsibility has to be drawn, giving away any allowance is a dangerous operation and we don't label it as a vulnerability but a property of the system.

**Comment by zobront** [comment]
+1 to @GalloDaSballo arguments.

**Comment by trust1995** [comment]
Agree that likelihood is very low. If we factor that in against potential high impact of lost user funds, and that only approved spender may abuse, it is between L and M. We can refer to the C4 risk [criterea](https://docs.code4rena.com/awarding/judging-criteria#estimating-risk)

QA (Quality Assurance) Includes both Non-critical (code style, clarity, syntax, versioning, off-chain monitoring (events, etc) and Low risk (e.g. assets are not at risk: state handling, function incorrect as to spec, issues with comments). Excludes Gas optimizations, which are submitted and judged separately.

2 — Med: Assets not at direct risk, but the function of the protocol or its availability could be impacted, or leak value with a hypothetical attack path with stated assumptions, but external requirements.

Which one sounds more like the issue at hand?

**Comment by GalloDaSballo** [comment]
> Agree that likelihood is very low. If we factor that in against potential high impact of lost user funds, and that only approved spender may abuse, it is between L and M. We can refer to the C4 risk [criterea](https://docs.code4rena.com/awarding/judging-criteria#estimating-risk)
> 
> QA (Quality Assurance) Includes both Non-critical (code style, clarity, syntax, versioning, off-chain monitoring (events, etc) and Low risk (e.g. assets are not at risk: state handling, function incorrect as to spec, issues with comments). Excludes Gas optimizations, which are submitted and judged separately.
> 
> 2 — Med: Assets not at direct risk, but the function of the protocol or its availability could be impacted, or leak value with a hypothetical attack path with stated assumptions, but external requirements.
> 
> Which one sounds more like the issue at hand?

Per the Rules, it only affects:
- one user
- non-zero to non-zero allowance
- Malcious target (this normally would be a given, but this argument holds only if you assume (to my disagreement) that the receiver cannot be trusted with the allowance)
- The attacker has X time to do the attack
- The caller doesn't prevent a front-run of the check
- The caller doesn't perform the intended operation (reset to 0)

Because those are layers and layers of conditionality, and because `approve` is the standard for ERC20s I believe Informational to be the only appropriate severity.

A Low severity implies this can happen with low Likelyhood but with a specific user mistake, this requires multiple layers of recklessness

A Low severity also implies that this can happen reliably, while the given risk can only happen if all those conditions align, and when they do, it's a direct outcome of misinterpretation of how a Public Mempool Blockchain system works and arguably a feature of it

**Comment by GalloDaSballo** [comment]
We need a malicious target as additional conditionality to give strength to your argument otherwise we would default to saying. "Any allowance is risky," while your argument directly implies that the marginal (the delta of allowance) is the part at risk

**Comment by trust1995** [comment]
I can get behind that reasoning.

**Comment by gpersoon** [comment]
I would say, the probability of it going wrong is low (you have to time it carefully and the receiver of the funds has to be malicious) and the impact usually is low to medium (unless maybe large amounts are being transferred).
Combining would lead to a low risk issue.
With solution of increase/decrease allowance.

**Comment by GalloDaSballo** [comment]
> I would say, the probability of it going wrong is low (you have to time it carefully and the receiver of the funds has to be malicious) and the impact usually is low to medium (unless maybe large amounts are being transferred). Combining would lead to a low risk issue. With solution of increase/decrease allowance.

Would like to ask you more precisely what makes you think is at risk wrt allowance, and why the marginal allowance would be at Low Risk vs the entire allowance (assuming a malicious target is given the allowance)

**Comment by HickupHH3** [comment]
I'd be more inclined to give it a medium severity if there's been an instance of the race condition being exploited. In the years I've been in crypto, I've not seen this to ever been a problem or part of an actual exploit. 

As @0xean pointed out, most users tend to give infinite approvals anyway.

+1 to NC, at the very most low.

**Comment by MrToph** [comment]
I agree with your first post, I never really saw this as an issue. Imo, it's a failure of Alice to think that any existing approval won't be spent before her new approve transaction is mined.

**Comment by GalloDaSballo** [comment]
I would like to shoutout @alcueca for being vindicated by the recent changes introduced by OZ

<img width="806" alt="Screenshot 2023-09-13 at 15 02 06" src="https://github.com/code-423n4/org/assets/13383782/4520281d-e3f6-4c97-8398-33d1bbae8242">

Original discussion:
https://github.com/code-423n4/2022-01-yield-findings/issues/6
